File: DSA/akashyadav_590014126/Day_40_Question_3.py
class max_stack:
    def __init__(self):
        self.main=[]
        self.maxes=[]
    def push(self,x):
        self.main.append(x)
        if not self.maxes or x>=self.maxes[-1]:
            self.maxes.append(x)
    def pop(self):
        if not self.main:
            return None
        val=self.main.pop()
        if val == self.maxes[-1]:
            self.maxes.pop()
        return val
    def getMax(self):
        if not self.maxes:
            return None
        return self.maxes[-1]

File: DSA/SAMAGRA_590014891/Day_40_Question_3.c
#include <stdio.h>
#include <limits.h>

#define MAX 10000

typedef struct {
    int stack[MAX], max_stack[MAX], top, max_top;
} MaxStack;

void init(MaxStack* s) {
    s->top = s->max_top = -1;
}

void push(MaxStack* s, int val) {
    s->stack[++s->top] = val;
    if (s->max_top == -1 || val >= s->max_stack[s->max_top])
        s->max_stack[++s->max_top] = val;
}

int pop(MaxStack* s) {
    if (s->top == -1) return INT_MIN;
    int val = s->stack[s->top--];
    if (val == s->max_stack[s->max_top]) s->max_top--;
    return val;
}

int getMax(MaxStack* s) {
    if (s->max_top == -1) return INT_MIN;
    return s->max_stack[s->max_top];
}

int main() {
    MaxStack s;
    init(&s);
    int q;
    scanf("%d", &q);
    while (q--) {
        char op[10];
        int x;
        scanf("%s", op);
        if (op[0] == 'p' && op[1] == 'u') {
            scanf("%d", &x);
            push(&s, x);
        } else if (op[0] == 'p') {
            pop(&s);
        } else if (op[0] == 'g') {
            printf("%d\n", getMax(&s));
        }
    }
    return 0;
}
